### PR TITLE
Fixes #18429 - Add foreman API support for trends

### PR DESCRIPTION
**Pull request by shiramax**

### CODE REVIEW

**Comment by ohadlevy** [comment]
@shiramax can you please add tests for the API endpoints? thanks.

**Comment by shiramax** [comment]
@ohadlevy , 
I added the tests, thanks.

**Comment by shiramax** [comment]
[test foreman]

**Comment by shiramax** [comment]
[test]

**Comment by shiramax** [comment]
[test foreman]

File: app/registries/foreman/access_permissions.rb [hunk 1]
```diff
@@ -791,11 +791,16 @@
...
-    map.permission :create_trends,  {:trends => [:new, :create]}
-    map.permission :edit_trends,    {:trends => [:edit, :update]}
-    map.permission :destroy_trends, {:trends => [:destroy]}
-    map.permission :update_trends,  {:trends => [:count]}
+    map.permission :view_trends,    {:trends => [:index, :show],
+                                     :"api/v2/trends" => [:index, :show]}
+    map.permission :create_trends,  {:trends => [:new, :create],
+                                      :"api/v2/trends" => [:new, :create]}
+    map.permission :edit_trends,    {:trends => [:edit, :update],
+                                     :"api/v2/trends" => [:edit, :update]}
```
Comment (ares) [inline app/registries/foreman/access_permissions.rb]: there's no edit in API I guess

File: app/registries/foreman/access_permissions.rb [hunk 2]
```diff
@@ -791,11 +791,16 @@
...
+    map.permission :destroy_trends, {:trends => [:destroy],
+                                     :"api/v2/trends" => [:destroy]}
+    map.permission :update_trends,  {:trends => [:count],
+                                     :"api/v2/trends" => [:count]}
```
Comment (ares) [inline app/registries/foreman/access_permissions.rb]: there's no count action in trends API controller if I'm not mistaken

File: app/models/trend.rb [hunk 3]
```diff
@@ -6,8 +6,10 @@ class Trend < ApplicationRecord
   belongs_to :trendable, :polymorphic => true
   has_many :trend_counters, :dependent => :destroy
 
+  scoped_search :on => [:trendable_type]
```
Comment (ares) [inline app/models/trend.rb]: :trendable_type does not have to be specified as an array if that's single value

**Review by ares — CHANGES_REQUESTED** [review]
I haven't tested it but leaving few comments that should be addressed

**Comment by shiramax** [comment]
@ares , 
Done, thanks!

**Comment by shiramax** [comment]
[test foreman]

**Comment by shiramax** [comment]
@ShimShtein

File: app/models/trend.rb [hunk 4]
```diff
@@ -6,8 +6,10 @@ class Trend < ApplicationRecord
   belongs_to :trendable, :polymorphic => true
   has_many :trend_counters, :dependent => :destroy
 
+  scoped_search :on => :trendable_type
```
Comment (ShimShtein) [inline app/models/trend.rb]: Why did you add this? It's not necessary for API.

File: app/models/trend.rb [hunk 5]
```diff
@@ -6,8 +6,10 @@ class Trend < ApplicationRecord
...
   scope :has_value, -> { where('fact_value IS NOT NULL').order("fact_value") }
   scope :types, -> { where(:fact_value => nil) }
+  scope :trend_source, ->  { where('fact_value IS NULL').order("trendable_type")}
```
Comment (ShimShtein) [inline app/models/trend.rb]: You can change it to 
``` ruby
scope :trend_source, ->  { where(:fact_value => nil).order(:trendable_type)}
```

File: app/controllers/api/v2/trends_controller.rb [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        end
+      end
+
+      api :DELETE, "/trends/:id/", N_("Delete a trend counter")
+      param :id, :identifier, :required => true
+      def destroy
+        process_response @trend.destroy
+      end
+
+      def find_resource
```
Comment (ShimShtein) [inline app/controllers/api/v2/trends_controller.rb]: It would be better to redefine `resource_scope`:

``` ruby
  def resource_scope(options = {})
    @resource_scope ||= scope_for(Trend.trend_source, options)
  end
```

File: test/controllers/api/v2/trends_controller_test.rb [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    @fact_trend = FactoryBot.create(:fact_trend, :trendable_type => 'FactName', :fact_name => 'kernelversion')
+    @fact_trend_valid_attrs = { :trendable_type => 'FactName', :trendable_id => @fact_os.fact_name_id.to_s }
+  end
+
+  test "should get index" do
+    get :index, { }
+    assert_response :success
+    assert_not_nil assigns(:trends)
+    trends = ActiveSupport::JSON.decode(@response.body)
+    assert !trends.empty?
```
Comment (ShimShtein) [inline test/controllers/api/v2/trends_controller_test.rb]: Can you add assertions about output layout please?

File: test/controllers/api/v2/trends_controller_test.rb [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+  end
+
+  test "should create foreman trend" do
+    assert_difference('Trend.trend_source.count',1) do
```
Comment (ShimShtein) [inline test/controllers/api/v2/trends_controller_test.rb]: You should assert that the trend was created with proper type:
``` ruby
assert_difference('Trend.trend_source.where(:type => "ForemanTrend").count',1)
```

File: test/controllers/api/v2/trends_controller_test.rb [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+      post :create, { :trend => foreman_trend_valid_attrs}
+    end
+    assert_response :created
+  end
+
+  test "should create fact trend" do
+    assert_difference('Trend.trend_source.count') do
```
Comment (ShimShtein) [inline test/controllers/api/v2/trends_controller_test.rb]: You should assert that the trend was created with proper type:
``` ruby
assert_difference('Trend.trend_source.where(:type => "FactTrend").count',1)
```

File: test/controllers/api/v2/trends_controller_test.rb [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    end
+    assert_response :error
+  end
+
+  test "should show individual record" do
+    get :show, { :id => @foreman_trend.id.to_param }
+    assert_response :success
+    show_response = ActiveSupport::JSON.decode(@response.body)
+    assert !show_response.empty?
+  end
```
Comment (ShimShtein) [inline test/controllers/api/v2/trends_controller_test.rb]: Add assertion about JSON layout please.

**Review by ShimShtein — CHANGES_REQUESTED** [review]

**Comment by shiramax** [comment]
@ShimShtein done, please review.

File: app/controllers/api/v2/trends_controller.rb [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+      end
+
+      api :DELETE, "/trends/:id/", N_("Delete a trend counter")
+      param :id, :identifier, :required => true
+      def destroy
+        process_response @trend.destroy
+      end
+
+      def resource_scope_for_index(options = {})
+        resource_scope(options).paginate(paginate_options)
```
Comment (ShimShtein) [inline app/controllers/api/v2/trends_controller.rb]: Could you please comment why you had to overload this method?

File: app/models/trend.rb [hunk 12]
```diff
@@ -8,6 +8,7 @@ class Trend < ApplicationRecord
 
   scope :has_value, -> { where('fact_value IS NOT NULL').order("fact_value") }
   scope :types, -> { where(:fact_value => nil) }
+  scope :trend_source, ->  { where('fact_value' => nil ).order("trendable_type")}
```
Comment (ShimShtein) [inline app/models/trend.rb]: Why do we need the `order` statement?

File: test/controllers/api/v2/trends_controller_test.rb [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+    assert_not_nil assigns(:trends)
+    trends = ActiveSupport::JSON.decode(@response.body)
+    assert !trends.empty?
+  end
+
+  test "should create a valid foreman trend" do
+    assert_difference('Trend.trend_source.where(:type => "ForemanTrend").count',1) do
+      post :create, { :trend => foreman_trend_valid_attrs}
+      result = ActiveSupport::JSON.decode(@response.body)
+      assert_equal(result["trendable_type"], "Model")
```
Comment (ShimShtein) [inline test/controllers/api/v2/trends_controller_test.rb]: Please put the expectation before the actual result, `assert_equal` signature is:
> assert_equal( expected, actual, failure_message = nil )

**Review by ShimShtein — CHANGES_REQUESTED** [review]
A few comments and still waiting for assertions in `show` and `index` tests.

Something along the lines of:
``` ruby
def assert_trend(expected, actual)
  assert_equal(expected["trendable_type"], actual["trendable_type"])
  assert_equal(expected["fact_name"], actual["fact_name"])
  assert_equal(expected["type"], actual["type"])
end

test "should get index" do
# ...
assert_trend(trends.first, { expected_trend })
```

File: app/controllers/api/v2/trends_controller.rb (hunk 11 above)
Comment (shiramax) [reply app/controllers/api/v2/trends_controller.rb]: in the resource_scope_for_index, there is usage in search_for method, 
and I can't use it because trends don't have this option

**Comment by shiramax** [comment]
@ShimShtein , Done
Please review again.
Thanks.

File: app/controllers/api/v2/trends_controller.rb (hunk 11 above)
Comment (ShimShtein) [reply app/controllers/api/v2/trends_controller.rb]: > in the resource_scope_for_index, there is usage in search_for method,
and I can't use it because trends don't have this option

Please leave this comment in code.

File: test/controllers/api/v2/trends_controller_test.rb [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+    @fact_trend_valid_attrs = { :trendable_type => 'FactName', :trendable_id => @fact_os.fact_name_id.to_s }
+  end
+
+  test "should get index" do
+    get :index, params: { }
+    assert_response :success
+    assert_not_nil assigns(:trends)
+    trends = ActiveSupport::JSON.decode(@response.body)
+
+    assert !trends.empty?
```
Comment (ShimShtein) [inline test/controllers/api/v2/trends_controller_test.rb]: Please add some asserts about json structure:

``` ruby
assert_equals [nil, 'some_fact'].sort, trends['results'].map { |t| t['fact_name'] }.sort
...
```

File: app/views/api/v2/trends/base.json.rabl [hunk 15]
```diff
@@ -0,0 +1,3 @@
+object @trend
+
+attributes :id , :trendable_type, :fact_name, :type
```
Comment (ShimShtein) [inline app/views/api/v2/trends/base.json.rabl]: I am missing `:name` attribute. Is it intentional?

File: test/controllers/api/v2/trends_controller_test.rb [hunk 16]
```diff
@@ -0,0 +1,69 @@
+require 'test_helper'
+
+class Api::V2::TrendsControllerTest < ActionController::TestCase
+  foreman_trend_valid_attrs = { :trendable_type => 'Model' }
```
Comment (ShimShtein) [inline test/controllers/api/v2/trends_controller_test.rb]: Just for the sake of uniformity, could you place this line inside `setup` block as class variable just as you do with `@fact_trend_valid_attrs`?

File: test/controllers/api/v2/trends_controller_test.rb [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+      post :create, params: { :trend => @fact_trend_valid_attrs }
+      result = ActiveSupport::JSON.decode(@response.body)
+      assert_equal("FactName", result["trendable_type"])
+      assert_equal("operatingsystem", result["fact_name"])
+      assert_equal("FactTrend", result["type"])
+    end
+    assert_response :created
+  end
+
+  test "should not create invalid tredns" do
```
Comment (ShimShtein) [inline test/controllers/api/v2/trends_controller_test.rb]: Typo `trends`

File: test/controllers/api/v2/trends_controller_test.rb [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+      post :create, params: { :trend => foreman_trend_invalid_attrs }
+    end
+    assert_response :error
+  end
+
+  test "should show individual record" do
+    get :show, params: { :id => @foreman_trend.id.to_param }
+    assert_response :success
+    show_response = ActiveSupport::JSON.decode(@response.body)
+    assert !show_response.empty?
```
Comment (ShimShtein) [inline test/controllers/api/v2/trends_controller_test.rb]: Please add asserts about JSON structure:

``` ruby
assert_equals @foreman_trend.trendable_type, show_response['trendable_type']
...
```

**Review by ShimShtein — CHANGES_REQUESTED** [review]
A few more tests plus small changes to the code and we are there!

**Comment by shiramax** [comment]
@ShimShtein ,
Done, Thanks.

**Comment by ShimShtein** [comment]
@shiramax, take a look at codeclimate's error. This time it is right, could you please refactor the code to merge both similar lines?

**Comment by shiramax** [comment]
@ShimShtein  Done

**Comment by shiramax** [comment]
@ShimShtein Done, thanks.

File: app/controllers/concerns/foreman/controller/parameters/trend.rb [hunk 19]
```diff
@@ -15,4 +15,14 @@ def trend_params_filter
   def trend_params
     self.class.trend_params_filter.filter_params(params, parameter_filter_context)
   end
+
+  def create_trend
```
Comment (ShimShtein) [inline app/controllers/concerns/foreman/controller/parameters/trend.rb]: This is nicer than code duplication we had, but parameter handling is not exactly the right place for this code.

I think it would be better to split this code to multiple places:
Trend creation should be dealt in factory method inside common model class that we have - `Trend`:

``` ruby
class Trend < ApplicationRecord
  def self.build_trend(trend_params = {})
    trend_params[:trendable_type] == 'FactName' ? FactTrend.new(trend_params) : ForemanTrend.new(trend_params)
  end
end
```

The rest of the method should be in a concern that would be consumed by both API and UI controllers:

``` ruby
# new file: app/controllers/concerns/foreman/controller/trends.rb
module Foreman::Controller::Trends
  def create_trend
    @trend = Trend.build_trend(trend_params)
    if @trend.save
      process_success
    else
      process_resource_error
    end
  end
end
```
Now both controllers can use this newly created concern:
``` ruby
class TrendsController < V2::BaseController
  include Foreman::Controller::Parameters::Trend
  include Foreman::Controller::Trends
  # ...

  api :POST, "/trends/", N_("Create a trend counter")
  def create
     create_trend
  end
end

# in the UI controller
class TrendsController < ApplicationController
  include Foreman::Controller::Parameters::Trend
  include Foreman::Controller::Trends
  
  #...

  def create
     create_trend
  end
end
```

**Comment by shiramax** [comment]
test foreman

**Review by ShimShtein — APPROVED** [review]

**Comment by ShimShtein** [comment]
LGTM, if @ares approves - I'll merge.

File: app/controllers/api/v2/trends_controller.rb [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+      end
+
+      api :GET, "/trends/:id/", N_("Show a trend")
+      param :trend_id, :identifier, :required => true
+      def show
+      end
+
+      api :POST, "/trends/", N_("Create a trend counter")
+      def create
+        @trend = Trend.build_trend(trend_params)
```
Comment (ares) [inline app/controllers/api/v2/trends_controller.rb]: `trend_params` should be more strict about what values we require, when I don't send trendable type, I receive a following trace

```
2018-01-16T08:26:58 962c054f [app] [W] Action failed
 | NoMethodError: undefined method `constantize' for nil:NilClass
 | /home/ares/Projekty/Zdrojaky/foreman/app/models/trends/foreman_trend.rb:17:in `create_values'
 | /home/ares/Projekty/Zdrojaky/foreman/app/models/trends/foreman_trend.rb:13:in `create_values'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activesupport-5.1.4/lib/active_support/callbacks.rb:413:in `block in make_lambda'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activesupport-5.1.4/lib/active_support/callbacks.rb:235:in `block in halting_and_conditional'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activesupport-5.1.4/lib/active_support/callbacks.rb:511:in `block in invoke_after'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activesupport-5.1.4/lib/active_support/callbacks.rb:511:in `each'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activesupport-5.1.4/lib/active_support/callbacks.rb:511:in `invoke_after'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activesupport-5.1.4/lib/active_support/callbacks.rb:132:in `run_callbacks'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activesupport-5.1.4/lib/active_support/callbacks.rb:827:in `_run_save_callbacks'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.1.4/lib/active_record/callbacks.rb:336:in `create_or_update'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.1.4/lib/active_record/persistence.rb:129:in `save'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.1.4/lib/active_record/validations.rb:44:in `save'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.1.4/lib/active_record/attribute_methods/dirty.rb:35:in `save'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.1.4/lib/active_record/transactions.rb:308:in `block (2 levels) in save'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.1.4/lib/active_record/transactions.rb:384:in `block in with_transaction_returning_status'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.1.4/lib/active_record/connection_adapters/abstract/database_statements.rb:235:in `block in transaction'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.1.4/lib/active_record/connection_adapters/abstract/transaction.rb:194:in `block in within_new_transaction'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/2.4.0/monitor.rb:214:in `mon_synchronize'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.1.4/lib/active_record/connection_adapters/abstract/transaction.rb:191:in `within_new_transaction'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.1.4/lib/active_record/connection_adapters/abstract/database_statements.rb:235:in `transaction'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.1.4/lib/active_record/transactions.rb:210:in `transaction'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.1.4/lib/active_record/transactions.rb:381:in `with_transaction_returning_status'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.1.4/lib/active_record/transactions.rb:308:in `block in save'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.1.4/lib/active_record/transactions.rb:323:in `rollback_active_record_state!'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.1.4/lib/active_record/transactions.rb:307:in `save'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.1.4/lib/active_record/suppressor.rb:42:in `save'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/bullet-5.7.0/lib/bullet/active_record5.rb:5:in `save'
 | /home/ares/Projekty/Zdrojaky/foreman/app/controllers/api/v2/trends_controller.rb:21:in `create'
```

perhaps the fix should go to `foreman_trend.rb:17:in `create_values'`

File: app/controllers/api/v2/trends_controller.rb [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+        @trends = resource_scope_for_index
+      end
+
+      api :GET, "/trends/:id/", N_("Show a trend")
+      param :trend_id, :identifier, :required => true
+      def show
+      end
+
+      api :POST, "/trends/", N_("Create a trend counter")
+      def create
```
Comment (ares) [inline app/controllers/api/v2/trends_controller.rb]: this and many other actions seems to be missing params definition, see e.g. https://github.com/theforeman/foreman/blob/develop/app/controllers/api/v2/domains_controller.rb#L42-L70 for example how to describe params

File: app/views/api/v2/trends/base.json.rabl [hunk 22]
```diff
@@ -0,0 +1,3 @@
+object @trend
+
+attributes :id, :trendable_type, :fact_name, :type, :name
```
Comment (ares) [inline app/views/api/v2/trends/base.json.rabl]: seems to be missing `trendable_id`

**Review by ares — DISMISSED** [review]
I left few comments that I think need to be fixed, also it seems update API endpoint is missing, is that intentional? The core of the functionality seems to work fine also under non-admin user :+1:

File: app/controllers/api/v2/trends_controller.rb (hunk 21 above)
Comment (ares) [reply app/controllers/api/v2/trends_controller.rb]: to reproduced, do this

```
curl -H "Accept:application/json,version=2" -H "Content-Type:application/json" -u admin:changeme -k https://foreman.example.tst/api/v2/trends -X POST -d '{}'
```

**Comment by iNecas** [comment]
There are still some comments that seem they need to be addressed, @shiramax mind looking?

**Comment by shiramax** [comment]
@iNecas @ares 
Done, thanks.

**Review by ares — COMMENTED** [review]
I will not get to rereview soon but if my comments were addressed, I'm fine with merging :-)

**Comment by ShimShtein** [comment]
[test]

**Comment by ShimShtein** [comment]
[test foreman]

**Comment by ShimShtein** [comment]
@shiramax, it looks like your PR needs a rebase.

**Comment by ShimShtein** [comment]
[test foreman]

File: app/controllers/api/v2/trends_controller.rb [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+        @trends = resource_scope_for_index
+      end
+
+      api :GET, "/trends/:id/", N_("Show a trend")
+      param :trend_id, :identifier, :required => true
+      def show
+      end
+
+      api :POST, "/trends/", N_("Create a trend counter")
+      param :trendable_type, String,  :required => true
```
Comment (ShimShtein) [inline app/controllers/api/v2/trends_controller.rb:19]: We should specify which params are we expecting here. I assume it's not only `trendable_type`

**Review by ShimShtein — CHANGES_REQUESTED** [review]
I see we allow `:type` parameter as part of `POST` request. Is it intentional?

**Comment by shiramax** [comment]
[test katello]

**Comment by shiramax** [comment]
@ShimShtein please review :)

File: app/models/trend.rb [hunk 24]
```diff
@@ -8,11 +8,16 @@ class Trend < ApplicationRecord
 
   scope :has_value, -> { where('fact_value IS NOT NULL').order("fact_value") }
   scope :types, -> { where(:fact_value => nil) }
+  scope :trend_source, -> { where('fact_value' => nil) }
```
Comment (ShimShtein) [inline app/models/trend.rb]: Why do we have `:trend_source` and `:types`. It's the same scope, isn't it?

**Review by ShimShtein — DISMISSED** [review]
Just a small question and we are good to go

**Comment by mmoll** [comment]
@shiramax please fix the rubocop violations, then it should be good to merge :green_heart:

**Review by mmoll — APPROVED** [review]

**Comment by mmoll** [comment]
merged, thanks @shiramax!
